crates/accelerate/src/sparse_pauli_op.rs

```diff
@@ -86,6 +132,199 @@ pub struct ZXPaulis {
+            x: self.x.bind(py).try_readonly().ok()?,
+            z: self.z.bind(py).try_readonly().ok()?,
+            phases: self.phases.bind(py).try_readonly().ok()?,
+            coeffs: self.coeffs.bind(py).try_readonly().ok()?,
+        })
+    }
+}
+
+/// Intermediate structure that represents readonly views onto the Python-space sparse Pauli data.
+/// This is used in the chained methods so that the syntactical temporary lifetime extension can
+/// occur; we can't have the readonly array temporaries only live within a method that returns
+/// [ZXPaulisView], because otherwise the lifetimes of the [PyReadonlyArray] elements will be too
+/// short.
+pub struct ZXPaulisReadonly<'a> {
+    x: PyReadonlyArray2<'a, bool>,
+    z: PyReadonlyArray2<'a, bool>,
+    phases: PyReadonlyArray1<'a, u8>,
+    coeffs: PyReadonlyArray1<'a, Complex64>,
+}
+
+impl ZXPaulisReadonly<'_> {
+    /// Get a [ndarray] view of the data of these [rust-numpy] objects.
+    fn as_array(&self) -> ZXPaulisView {
+        ZXPaulisView {
+            x: self.x.as_array(),
+            z: self.z.as_array(),
+            phases: self.phases.as_array(),
+            coeffs: self.coeffs.as_array(),
+        }
+    }
+}
+
+/// Intermediate structure that represents [ndarray] views onto the Python-space sparse Pauli data
+/// in the ZX convention.  This can be used directly by Rust methods if desired, or bit-packed into
+/// a matrix-representation format [MatrixCompressedPaulis] using the [compress] method.
+pub struct ZXPaulisView<'py> {
+    x: ArrayView2<'py, bool>,
+    z: ArrayView2<'py, bool>,
+    phases: ArrayView1<'py, u8>,
+    coeffs: ArrayView1<'py, Complex64>,
+}
+
+impl<'py> ZXPaulisView<'py> {
+    /// The number of qubits this operator acts on.
+    pub fn num_qubits(&self) -> usize {
+        self.x.shape()[1]
+    }
+
+    /// Convert the ZX representation into a bitpacked internal representation.  See the
+    /// documentation of [MatrixCompressedPaulis] for details of the changes to the Pauli
+    /// convention and representation.
+    pub fn matrix_compress(&self) -> PyResult<MatrixCompressedPaulis> {
+        let num_qubits = self.num_qubits();
+        // This is obviously way too big for a dense operator, and SciPy limits us to using `i64`
+        // for the index and indptr types, so (except for some synthetic cases), it's not possible
+        // for us to work with a larger matrix than this.
+        if num_qubits > 63 {
+            return Err(PyValueError::new_err(format!(
+                "{num_qubits} is too many qubits to convert to a matrix"
+            )));
+        }
+        if num_qubits == 0 {
+            return Ok(MatrixCompressedPaulis {
+                num_qubits: 0,
+                x_like: Vec::new(),
+                z_like: Vec::new(),
+                coeffs: self.coeffs.to_vec(),
+            });
+        }
+        let x_like = pack_bits(self.x).unwrap();
+        let z_like = pack_bits(self.z).unwrap();
```

Comment:
`unwrap` failing here would have been an internal logic error, because `ZXPaulis` validates that `x` and `z` have the same shape on construction from Python space, then the start of `ZXPaulisView::matrix_compress` validates that the number of qubits will fit in a `u64` after the packing, which is the error condition for this (and elevates that to a `ValueError`).

I switched these over to be `expect` with an error message explaining where the internal-logic error would come in, but can switch them to full Python exceptions if you think it's better (though it _should_ be impossible to trigger them, even from Rust space, using the public interface).

Reply:
Hmm, if you're saying that the input to `pack_bits` will always fit in 64 bits because of earlier assertions than I would probably just remove the size check and result return from the function then.

Reply:
Whoops sorry, I got out-of-sync with your comments.  The "done in" is referring to my prior comment.

The `pack_bits` function could still fail if it were called from other valid places in Rust space using its regular public interface (and as we expand `SparsePauliOp` more in Rust space, we're more likely to use that), so I think it's right to return a `Result`.  It's just that in this particular case, we already validated the inputs so we know it should always work.